### PR TITLE
[develop] Use templates for METplus config files

### LINKED ISSUE #382: Separate out ensemble verification (gen-ens-prod and ensemble-stat)
Original issue opened by @michelleharrold migrated from regional_workflow repository: https://github.com/ufs-community/regional_workflow/issues/709

## Description
In METv10.1, the ensemble-stat tool has been split into two distinct tools: gen-ens-prod (new tool; creates ensemble products) and ensemble-stat (calculates ensemble stats). While METv10.1 supports backward compatibility of old ensemble-stat tool (product generation + ensemble verification), it would be good to pivot to new functionality/tools before it becomes deprecated.

## Solution
Breakout out current ensemble verification that uses MET ensemble-stat tool into gen-ens-prod tool and ensemble-stat tool.

### PR DESCRIPTION
**Pull request by gsketefian**
## DESCRIPTION OF CHANGES: 

This PR converts the METplus configuration files into Jinja templates form which the config files needed by the vx tasks can be generated.  These config files are placed in the experiment directory.  This prevents accidental changes to the METplus config files when multiple experiments are running simultaneously.  This PR also reduces the number of J-jobs and ex-scripts used by the vx tasks by combining ones that are similar.  Details:
1) Convert the METplus configuration files into Jinja templates.  In the process, to avoid repetition, combine multiple files for APCP hours greater than 01 (03, 06, and 24) into a single file having the string `_APCPgt01h` in its name.
2) Modify the ex-scripts so that they use the `fill_jinja_template.py` script to generate the METplus config files from the templates.
3) Combine similar j-jobs and ex-scripts into ones that can be used by multiple tasks, e.g. combine the ex-script that calls METplus's `GridStat` and `PointStat` tools for deterministic verification into one ex-script with the string `gridstat_or_pointstat` in its name.
4) Modify the ROCOTO template XML to use the new J-job and ex-script names and pass in several new variables to the vx tasks (`OBTYPE`, `ACCUM_HH` (if necessary), and `MET_TOOL`; also, remove `FHR`).
5) Add a new WE2E test (`MET_ensemble_verification_only_vx`) that tests use of staged obs and forecast files to run ensemble verification. 
6) Fix/clean up running of vx tasks in NCO mode.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## TESTS CONDUCTED: 

- [x] hera.intel
- [ ] orion.intel
- [ ] cheyenne.intel
- [ ] cheyenne.gnu
- [ ] gaea.intel
- [ ] jet.intel
- [ ] wcoss2.intel
- [ ] NOAA Cloud (indicate which platform)
- [ ] Jenkins
- [x] fundamental test suite
- [ ] comprehensive tests (specify *which* if a subset was used)

Ran the `fundamental` tests on Hera with intel (see below); all passed.
```
MET_verification
MET_verification_only_vx
community_ensemble_2mems
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp_regional_plot
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v16
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_RAP_suite_HRRR
grid_RRFS_CONUS_25km_ics_GSMGFS_lbcs_GSMGFS_suite_GFS_v15p2
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_HRRR
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_2017_gfdlmp_regional_plot
MET_ensemble_verification
MET_ensemble_verification_only_vx
community_ensemble_2mems_stoch
pregen_grid_orog_sfc_climo
```

## DOCUMENTATION:

Documentation is needed but will be added once the whole set of vx PRs has been merged (see Issue #630).

## ISSUE:
Resolves #382 

## CHECKLIST

- [x] My code follows the style guidelines in the Contributor's Guide
- [x] I have performed a self-review of my own code using the Code Reviewer's Guide
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes do not require updates to the documentation (explain).
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

### CODE REVIEW

**Comment by gsketefian** [comment]
@danielabdi-noaa This PR doesn't yet work in NCO mode, but I'm working on that now and will hopefully get it done this afternoon.

@christinaholtNOAA This is the PR I was referring to on Friday.  It has only a few minor (and duplicate) mods to the vx tasks in `FV3LAM_wflow.xml`.  Not sure where you're at with your PR.  If you get your PR in before mine, that's fine, I will incorporate the xml changes afterward.  Also, please let me and @MichaelLueken know if you'd like us to wait after yours is in before merging this PR; that should be fine.

@JeffBeck-NOAA @linlin-pan @willmayfield @michelleharrold In case you have time to review, that would be great.  Thanks.

**Review by JeffBeck-NOAA — APPROVED** [review]
Approved pending any necessary NCO mode changes.

File: jobs/JREGIONAL_RUN_MET_GENENSPROD_OR_ENSEMBLESTAT
Comment (mkavulich) [inline jobs/JREGIONAL_RUN_MET_GENENSPROD_OR_ENSEMBLESTAT]: All of this `_sc` vs `_pc` logic seems very complicated for something that doesn't seem to influence the code in any way; unless I'm missing something the "snake case" variables are only ever printed to log files. Is there some other reason I am missing?

**Comment by gsketefian** [comment]
@danielabdi-noaa @christinaholtNOAA @MichaelLueken @JeffBeck-NOAA Here's an update:
1) I fixed (and tested) NCO mode, in the process cleaning up where the vx output gets placed (output from some tasks were (I think inadvertantly) being placed in redundant CDATE-based subdirectories under `$COMOUT`).
2) I intentionally did not make any big changes to `FV3LAM_wflow.xml` to keep it easier to integrate into the new XML generation system of @christinaholtNOAA.
3) I retested all the fundamental tests on Hera.  I also ran the four vx tests (`MET_verification`, `MET_verification_only_vx`, `MET_ensemble_verification`, and `MET_ensemble_verification_only_vx`) in both `community` and `nco` modes, and they all passed.

So this PR is ready for the automated tests (and needs another approval).  Thanks.

**Comment by gsketefian** [comment]
@danielabdi-noaa @JeffBeck-NOAA @MichaelLueken If you're interested in the layout of the vx output in NCO mode, you can look on Hera under the directory
```
/scratch2/BMC/det/Gerard.Ketefian/UFS_CAM/DTC_ensemble_task_pythonize_nco/expt_dirs
```
in the subdirectories
```
nco_dirs.works01.MET_ensemble_verification_nco
nco_dirs.works01.MET_ensemble_verification_only_vx_nco
nco_dirs.works01.MET_verification_nco
nco_dirs.works01.MET_verification_only_vx_nco
```
Under these subdirectories, look in
```
com/rrfs/v1.0.0/rrfs.20190615/00/metout/
```

**Review by danielabdi-noaa — APPROVED** [review]
Thanks @gsketefian. Looks good to me!

File: parm/metplus/Pb2nc_obs.conf
Comment (MichaelLueken) [inline parm/metplus/Pb2nc_obs.conf]: Should prepbvur be prepbufr?

**Review by MichaelLueken — APPROVED** [review]
@gsketefian These changes look good to me!  I have noted a minor spelling issue in `parm/metplus/Pb2nc_obs.conf` (prepbuvr rather than prepbufr), which I have gone ahead and noted in my review.

Will go ahead and approve and submit automated Jenkins testing now.  Please note that Gaea is still down, so it might take some time for the tests to fully complete.

**Comment by MichaelLueken** [comment]
@gsketefian All tests have successfully passed, with the exception of Gaea, which is still down.

File: parm/metplus/Pb2nc_obs.conf
Comment (gsketefian) [reply parm/metplus/Pb2nc_obs.conf]: Yep!  Fixed it.

**Comment by mkavulich** [comment]
@gsketefian When I tried running the verification WE2E tests on Hera, they all showed failures in GridStat tasks for ACPC. Here is an example log file: /scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_683/expt_dirs/community/MET_verification/log/run_MET_GridStat_vx_APCP03h_2019061500.log

**Review by mkavulich — CHANGES_REQUESTED** [review]
@gsketefian When I tried running the verification WE2E tests on Hera, they all showed failures in GridStat tasks for ACPC. Here is an example log file: /scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_683/expt_dirs/community2/MET_verification_only_vx/log/run_MET_GridStat_vx_APCP03h_2019061500.log

**Comment by mkavulich** [comment]
Sorry about all the comments; I accidentally hit "submit" earlier on an earlier comment but I wanted to get my note about the failures before this was merged.

To go into further detail, I'm not sure why my tests are failing while yours seem to be succeeding. I started with a fresh clone of your branch on Hera (/scratch2/BMC/fv3lam/kavulich/UFS/workdir/PR_683/ufs-srweather-app/). It seems to be specifically the APCP03 and 06 GridStat tasks that are failing: the APCP01 GridStat tasks are successful.

@MichaelLueken Can you point me to the location on disk of the Jenkins test on Hera so I can see what the successful log files look like?

File: jobs/JREGIONAL_RUN_MET_GENENSPROD_OR_ENSEMBLESTAT
Comment (gsketefian) [reply jobs/JREGIONAL_RUN_MET_GENENSPROD_OR_ENSEMBLESTAT]: Yeah, it's left over from the time when some things were named "grid_stat_..." and others used "GridStat_...".  Let me take a look and see if I can simplify.

**Comment by gsketefian** [comment]
@mkavulich It failed because I had neglected to commit a new conf file (that I have on disk, so my tests succeed).  I just added it and hopefully it will work.

@MichaelLueken I'm also curious how the tests succeeded without that file...

File: jobs/JREGIONAL_RUN_MET_GENENSPROD_OR_ENSEMBLESTAT
Comment (gsketefian) [reply jobs/JREGIONAL_RUN_MET_GENENSPROD_OR_ENSEMBLESTAT]: @mkavulich I cleaned up the MET/METplus naming code, hopefully to your approval.  I removed the code in the J-jobs, and now the ex-scripts don't need arguments and get the variable `MET_TOOL` directly from the environment.  I added a function (`get_met_metplus_tool_name.sh`) that sets the names of the MET and METplus tools.  I want to keep both names still because they are both currently used; maybe we'll simplify the ex-scripts later and use only one of the names to refer to the tool, but let's leave that for another PR.  Also, I know we want to do all python scripts, but I haven't gotten a chance to do that yet (will do it in a future PR if someone else doesn't get to it before).  Let me know what you think.

Also, I reran the WE2E tests `MET_verification_only_vx` and `MET_ensemble_verification_only_vx`, and the both passed.

**Comment by gsketefian** [comment]
@mkavulich Did your manual NCO-mode tests on Hera succeed?

**Comment by MichaelLueken** [comment]
@mkavulich and @gsketefian The Jenkins tests don't run on Hera currently.  Currently, automated tests only run on Cheyenne, Jet, Gaea, and Orion.  While new allocations are in place for Hera, Jet, and Gaea, I'm unsure whether these new allocations have been propagated through Jenkins.

**Comment by mkavulich** [comment]
@gsketefian The tests on Hera have passed on your latest changes. But RE: my comments about MET tool names, I don't think you understood my point...I was looking to simplify some logic, but your latest commits have only *increased* the complexity of the logic. It's clever, sure, but it seems like completely unnecessary complexity. Why are we maintaining a whole bunch of logic (and now even a special function) to maintain two different variable names for each MET tool when one of the names is never even used except for in print statements? ush/get_met_metplus_tool_name.sh is 124 lines that does nothing but take one string and return two different strings....but one of those strings (`met_tool_name`) never gets used except in print statements! Can we not just use that second string (`metplus_tool_name`) in print statements and greatly simplify the logic? 

Again, maybe there's a reason these additional variables are needed that I'm missing, but I don't see it.

**Comment by gsketefian** [comment]
@mkavulich I'm anticipating that we may need to refer to MET vs. METplus versions of the tool in the ex-scripts at some point.  Before this PR, both were used in various places.  I went through all the config files and converted as much of the MET tool references to METplus references.  I'm getting to what you're suggesting (getting rid of all MET tool references), but that's something we should discuss in the DTC vx group.  We may at some point need to call MET without going through METplus.  I tried to reduce everything to that one function to still keep the capability but have it all in one place (so it's easy to remove in the future if we want).  I prefer to keep it and remove after a discussion, but if you feel very strongly then I'll remove it.

**Comment by mkavulich** [comment]
@gsketefian I'm fine with that reasoning, just wanted to make sure we weren't incurring technical debt without a reason.

**Review by mkavulich — APPROVED** [review]

**Review by panll — APPROVED** [review]

**Comment by JeffBeck-NOAA** [comment]
@MichaelLueken, I think this PR is ready.  The only problem is that @gsketefian is on leave this week and there is a minor conflict with the hera.yaml machine file (change in default paths).  Are you comfortable with me resolving this conflict and then proceeding with the merge?

**Comment by MichaelLueken** [comment]
@JeffBeck-NOAA I certainly don't mind if you went ahead and corrected the conflict.  The issue is that PRs that change the FV3LAM_wflow.xml file are on hold until @christinaholtNOAA's PR #676 has been merged.

@christinaholtNOAA Since both this PR and @chan-hoo's PR #682 have been approved and automated tests have been run and passed, can we move forward with these two PRs?  Unfortunately, it would mean needing to apply these two updates to your PR before your work can be merged.  Thank you very much for your time and consideration.

**Comment by chan-hoo** [comment]
@MichaelLueken, please merge my PR https://github.com/ufs-community/ufs-srweather-app/pull/682 before the PR https://github.com/ufs-community/ufs-srweather-app/pull/676 if it passed the tests :) I am going to create the `production/aqmv7` branch for production of AQMv7 once my PR is merged. It (without the huge change) will make my life much easier :)

**Comment by MichaelLueken** [comment]
@JeffBeck-NOAA Speaking with @christinaholtNOAA, the final changes should be in place and we are hoping that the Jenkins tests will be submitted today on PR #676.  So long as there are no failures with the Jenkins tests and manual testing on Jet, then her PR will be merged either tomorrow or first thing Friday.  If there are issues, then I will be able to move forward with the changes in this PR.  Thanks!

**Comment by JeffBeck-NOAA** [comment]
@MichaelLueken, thanks.  I've resolved the conflicts for this PR and merged the commit.

**Comment by gsketefian** [comment]
@MichaelLueken I updated this PR and reran (on Hera) the four verification tests:
```
MET_verification_only_vx
MET_ensemble_verification_only_vx
MET_verification
MET_ensemble_verification
```
In addition, I ran the fundamental tests, which currently are:
```
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_RAP_suite_RRFS_v1beta
nco_grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_timeoffset_suite_GFS_v16
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15p2
grid_RRFS_CONUS_25km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v17_p8
grid_RRFS_CONUScompact_25km_ics_HRRR_lbcs_HRRR_suite_HRRR
grid_SUBCONUS_Ind_3km_ics_HRRR_lbcs_RAP_suite_WoFS_v0
grid_RRFS_CONUS_25km_ics_NAM_lbcs_NAM_suite_GFS_v16
```
All tests were successful.  I think this PR is ready for the Jenkins tests.

**Comment by gsketefian** [comment]
@MichaelLueken I see the error message "This commit cannot be buiild".  I'd like the see the logs, but clicking on "Details" doesn't take me there (browser hangs).  Are the logs not public?

**Comment by MichaelLueken** [comment]
@gsketefian Unfortunately, the Jenkins logs are no longer public due to security concerns.  I'll look through the failures now, but the due to the upgrading of PrgEnv on Gaea this week, Gaea is having issues.  I'll let you know if there are any failures that will need to be addressed before this work can be merged.

**Comment by MichaelLueken** [comment]
@gsketefian 

On Cheyenne GNU, the `nco_grid_RRFS_CONUS_3km_ics_FV3GFS_lbcs_FV3GFS_suite_GFS_v15_thompson_mynn_lam3km` test failed intermittently in the post.  This is an ongoing issue that has been seen in all PRs since the test was added to the coverage suite and might be due to issue #652.  This will not keep your PR from being merged.

On Hera GNU, both the `get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2019061200` and `get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS` tests failed.  The `get_from_NOMADS_ics_FV3GFS_lbcs_FV3GFS` test has been failing since it was introduced into the coverage tests.  A rerun of the `get_from_HPSS_ics_FV3GFS_lbcs_FV3GFS_fmt_nemsio_2019061200` test showed that it passes.  Can safely move forward.

On Hera Intel, the `grid_RRFS_CONUS_13km_ics_FV3GFS_lbcs_FV3GFS_suite_RRFS_v1beta` test failed in the run_fcst task.  This is known issue (#731).

Since the failures are expected and the rerun of the one unexpected failure successfully passed, I will now move forward with merging this work.

**Comment by gsketefian** [comment]
@MichaelLueken Great, thank you!

**Pull request merged**
